lib/enum.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+# !/home/eduardo/.rvm/rubies/ruby-2.6.3/bin/ruby
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    to_a.length.times { |i| yield(to_a[i]) }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    to_a.length.times { |i| yield(to_a[i], i) }
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    arr = []
+    my_each { |i| arr.push(i) if yield(i) }
+    arr
+  end
+
+  def my_all?(pattern = nil)
+    if block_given?
+      my_each { |i| return false unless yield(i) }
+    elsif pattern.nil?
+      my_each { |i| return false unless i }
+    else
+      my_each { |i| return false unless i =~ pattern }
+    end
+    true
+  end
+
+  def my_any?(pattern = nil)
```

Comment:
Should work with classes passed as arguments.  And work with regular pattern too